module/zfs/spa_misc.c

```diff
@@ -1760,11 +1759,21 @@ spa_freeze_txg(spa_t *spa)
  * block anyway.
  */
 uint64_t
-spa_get_worst_case_asize(spa_t *spa, uint64_t lsize)
+spa_get_worst_case_asize(spa_t *spa, objset_t *os, uint64_t lsize)
 {
+	ASSERT3U(spa->spa_worst_alloc, >, 0);
 	if (lsize == 0)
 		return (0);	/* No inflation needed */
-	return (MAX(lsize, 1 << spa->spa_max_ashift) * spa_asize_inflation);
+
+	uint64_t inflation = roundup(lsize, (1 << spa->spa_max_ashift) *
+	    spa->spa_worst_alloc);
```

Comment:
This seems suboptimal. An ashift=9 raidz3 with an ashift=13 mirror would inflate to 32KB. Since tvds know both their geometry and ashift, IMO `vdev_op_worst_alloc` should just the return the minimum allocation unit in bytes instead of a multiplier. Then `spa_worst_alloc` be the maximum of those.

Reply:
Before this patch, it would have been MAX(lsize, spa_max_ashift=13) * spa_asize_inflation=24, and the lsize would be multiplied by 24 (so 192KB) rather than rounded up to the nearly 32KB.

The issue is that at this point ZFS has not decided which vdev it is going to write the data to. So it has to assume the worst case. Even if we moved the multiplication by ashift to the `vdev_op_worst_alloc` it would result in the same calculation, because we'd use the worst case for any vdev in the pool.

Reply:
I mean do the shift in `vdev_op_worst_alloc`, like it is done in `vdev_draid_min_alloc`. Forget `spa_max_ashift`.

Reply:
Either way the calculation will still be exactly the same, since we don't know which VDEV we are going to use, it will be the worst of the results from `vdev_op_worst_alloc`, so it would still be raidz3=4 * ashift=13

It doesn't change the math